src/main/java/net/minecraftforge/common/ForgeHooks.java

```diff
@@ -1398,4 +1399,30 @@ public static void modifyAttributes()
             FORGE_ATTRIBUTES.put(k, newBuilder.build());
         });
     }
+
+    private static final Set<BlockEntity> LOADING_BLOCK_ENTITIES = Sets.newConcurrentHashSet();
+
+    @Deprecated
+    public static void enqueueLoadingBlockEntity(BlockEntity be)
+    {
+        LOADING_BLOCK_ENTITIES.add(be);
+    }
+
+    @Deprecated
+    public static void onLevelTick(Level level)
```

Comment:
Waiting all the way for the first tick just to make sure you can access the chunk from its level will probably work, but it seems a bit drastic.

Don't we just need to wait for the future to resolve?
Then we don't even need to keep track of the "loading" block entities. We can just iterate over the ones in the chunk.

Reply:
This really depends on wether we want to allow full access to the chunk in `IForgeBlockEntity#onLoad()`. In 1.16 for example this is not possible.
If we want to go the same route as 1.16, then the necessary change is incredibly simple and no defferal to the first tick is necessary.
If full access to the chunk is desired then there is (as far as I can see) no way around this whole thing. In theory I can guard against the deadlock created by accessing the chunk while it is loading. In practice this only works as long as the only chunk that is accessed by whatever you do in `onLoad()` is the one that is loading. As soon as you access another chunk (either by your `BlockEntity` being placed on a chunk border and doing something that causes a block update (like the test mod) or your code explicitly reaching into other chunks) you get another deadlock that can't be guarded against because that other chunk doesn't start loading until the current one is done.

Simply waiting for the future to resolve is also not an option because Mojang are not only using the futures for loading the chunk but also for retrieving a loaded chunk when it is not contained in their "cache array" holding the 4 last used chunks.

Reply:
> This really depends on wether we want to allow full access to the chunk in IForgeBlockEntity#onLoad(). In 1.16 for example this is not possible.

I think there is a discussion to be had on what the intention here is because clearly this is impacting simplicity somewhat. If it never worked like that in 1.16, I don't know if there are good reasons to change that. Not saying there aren't, but I'm not aware of what people typically do in their `onLoad` calls. All I'm personally looking for is for my model data to be loaded, which doesn't require the level. In fact, I don't even override the `onLoad` call at all.

---

I was typing up a whole story about a good place to add a hook when I decided to check about existing hooks. Would the `ChunkEvent.Load` event (or a hook in the same place) not suffice? That one gets called during promotion on the `ChunkMap` itself, so the mentioned cache doesn't play a role.

Reply:
If you need example of how other mods use onLoad, my use of onLoad function includes deferring the spawning of a fresh entity around my tile entity, through decreasing a counter in onLoad to allow some entities in my tick method to be spawned once that counter reaches 0.

I also send packets to update data for models in the onLoad method for another tile.

Not sure if this method is how we are meant to use it, but that is how i currently do so.

Reply:
> I was typing up a whole story about a good place to add a hook when I decided to check about existing hooks. Would the `ChunkEvent.Load` event (or a hook in the same place) not suffice? That one gets called during promotion on the `ChunkMap` itself, so the mentioned cache doesn't play a role.

This event is called in the same place as the Chunk method that calls my current "enqueue" hook. This still happens in the running future which means accessing the Chunk through the Level in the BlockEntity will cause a deadlock.
The main problem here is that the futures are blocking the main thread and the one spot where that blocking is done has no idea if the requested chunk is has to be loaded/generated or not. Writing this gave me an idea though...

Reply:
My impression was that at this point, the promotion has been done and actual future resolves to the chunk, but I am probably missing something. It's not easy to navigate this code.

> Writing this gave me an idea though...
I'd love to hear/see it.

Reply:
> My impression was that at this point, the promotion has been done and actual future resolves to the chunk, but I am probably missing something. It's not easy to navigate this code.

The problem is that anything that accesses a chunk needs to happen outside of the future because accessing a chunk creates a future that then is enqueued in the `BlockableEventLoop` because that is passed as the `Executor` to the future. Because that new future is enqueued after the running future, it will not complete before the one currently running and therefore block the running future.

> I'd love to hear/see it.

```
CompletableFuture<Either<ChunkAccess, ChunkHolder.ChunkLoadingFailure>> completablefuture = this.getChunkFutureMainThread(p_8360_, p_8361_, p_8362_, p_8363_);
boolean loaded = completablefuture.isDone(); //Check if the chunk was already loaded
this.mainThreadProcessor.managedBlock(completablefuture::isDone);
ChunkAccess chunkaccess1 = completablefuture.join().map((p_8406_) -> {...});
this.storeInCache(i, chunkaccess1, p_8362_);
if (!loaded) {
    //Iterate BEs to call onLoad()
}
return chunkaccess1;
```
This is the piece of code that I mentioned that blocks the main thread. The `CompletableFuture` gets the `mainThreadProcessor` (which is a `BlockableEventLoop`) as its `Executor`. The CF calls `execute()` on that which leads to enqueueing the CFs `Runnable` in the event loops queue. The call to `managedBlock()` then works through queued tasks until the `Predicate` it gets passed (`completablefuture::isDone` in this case) returns true. One of the queued tasks will be the chunk load task. Because of that queueing I am able to check if the CF is already done (which would indicate that this Chunk was already loaded earlier). This works perfectly fine in theory but in testing I found that there is another path where chunk loading can happen (which probably has to do with either the spawn chunks or the chunks the player is in, I have not found the specific code path yet) that completely bypasses the method this code block is in and therefore makes this approach useless.